nemo_aligner/models/nlp/gpt/gpt_steerlm_model.py

```diff
@@ -0,0 +1,357 @@
+                )
+                cp_size = parallel_state.get_context_parallel_world_size()
+                if validation_step and not self.cfg.data.get("validation_drop_last", True):
+                    num_valid_tokens_in_ub = batch["num_valid_tokens_in_ub"]
+                    if loss_for_ub.isnan():
+                        assert batch["loss_mask"].count_nonzero() == 0, "Got NaN loss with non-empty input"
+                        loss_sum_for_ub = torch.zeros_like(num_valid_tokens_in_ub)
+                    else:
+                        loss_sum_for_ub = num_valid_tokens_in_ub * loss_for_ub
+
+                    loss_sum_and_ub_size_all_gpu = torch.cat(
+                        [
+                            loss_sum_for_ub.clone().detach().view(1),
+                            torch.tensor([num_valid_tokens_in_ub]).cuda().clone().detach(),
+                        ]
+                    )
+                    # Could potentially reduce num_valid_samples_in_microbatch and use that to aggregate instead of len(self._validation_ds)
+                    torch.distributed.all_reduce(
+                        loss_sum_and_ub_size_all_gpu, group=parallel_state.get_data_parallel_group()
+                    )
+                    return loss_for_ub * cp_size, {"loss_sum_and_ub_size": loss_sum_and_ub_size_all_gpu}
+                else:
+                    reduced_loss = average_losses_across_data_parallel_group([loss_for_ub])
+                    return loss_for_ub * cp_size, {"avg": reduced_loss}
+
+            return output_tensor, loss_func
+
+        return fwd_output_and_loss_func
+
+    def get_loss_and_metrics(self, batch, forward_only):
+        """Take a data_iter which is an interator over the microbatches
+            and return loss as well as metrics
+        """
+        # first compute the baseline weights
+        set_sync_funcs(self, True)
+        distances = []
+
+        # calculate the baseline weights for each group of generated reponses
+        # set the batch size for log likelihood calculation
+        response_size = batch["num_responses"][0].item()
+        gbs, seq_length = batch["tokens"].shape
+        mbs = int(self.cfg.data.steerlm2_weight_micro_batch_size)
+        dp_size = 1
+        configure_batch_sizes(mbs=mbs, gbs=gbs, dp=dp_size)
+
+        group_batch = {k: v for k, v in batch.items() if isinstance(v, torch.Tensor)}
+
+        data_iter = get_iterator_k_split(group_batch, get_num_microbatches())
+
+        fwd_bwd_function = get_forward_backward_func()
+
+        # first get the forward nll loss for each of the microbatches
+        fwd_loss_fn = self.get_forward_output_only_for_weight_func()
+        losses_forward = fwd_bwd_function(
+            forward_step_func=fwd_loss_fn,
+            data_iterator=data_iter,
+            model=self.model,
+            num_microbatches=get_num_microbatches(),
```

Comment:
nit: instead of changing the global num microbatches/micro batch size you can directly give your desired value to mcore and it'll work

Reply:
done